File: sdks/node/src/online.ts
/**
 * OnlineClient — submit requests and wait for results from hub-router.
 * Fully async, non-blocking. Works with Express, Fastify, NestJS, etc.
 */
import {
  HubRouterError,
  QueueFullError,
  RequestNotFoundError,
  ResultTimeoutError,
} from "./errors.js";
import type {
  OnlineClientOptions,
  Result,
  SubmitResponse,
} from "./types.js";

const DEFAULT_LONG_POLL_TIMEOUT_MS = 30_000;
const DEFAULT_MAX_RETRIES = 10;

export class OnlineClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly longPollTimeoutMs: number;
  private readonly maxRetries: number;

  /**
   * @param baseUrl   hub-router base URL, e.g. "https://hub.example.com"
   * @param apiKey    X-Online-API-Key value
   * @param options   Optional overrides
   */
  constructor(
    baseUrl: string,
    apiKey: string,
    options: OnlineClientOptions = {}
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.longPollTimeoutMs =
      options.longPollTimeout ?? DEFAULT_LONG_POLL_TIMEOUT_MS;
    this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
  }

  /**
   * Enqueue a request and return its correlation ID.
   *
   * @throws {QueueFullError} if the queue is at capacity.
   */
  async submit(
    payload: unknown,
    headers: Record<string, string> = {}
  ): Promise<string> {
    const res = await fetch(`${this.baseUrl}/request`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Online-API-Key": this.apiKey,
      },
      body: JSON.stringify({ payload, headers }),
    });

    if (res.status === 503) {
      const data = (await res.json()) as { message?: string };
      throw new QueueFullError(data.message);
    }
    if (res.status !== 202) {
      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }

    const data = (await res.json()) as SubmitResponse;
    return data.id;
  }

  /**
   * Long-poll until the result for *requestId* is ready.
   *
   * Each poll hangs for up to `longPollTimeout` ms server-side.
   * A 204 response means the server timed out waiting — this client retries
   * automatically (up to `maxRetries` times).
   *
   * @throws {RequestNotFoundError} if the ID is unknown or expired.
   * @throws {ResultTimeoutError}   if `maxRetries` exhausted.
   */
  async waitResult(requestId: string, options?: { timeout?: number }): Promise<Result> {
    const timeoutMs = options?.timeout ?? this.longPollTimeoutMs;
    const timeoutSec = (timeoutMs / 1000).toFixed(1);
    const url = `${this.baseUrl}/result/${encodeURIComponent(requestId)}?timeout=${timeoutSec}s`;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const res = await fetch(url, {
        headers: { "X-Online-API-Key": this.apiKey },
        signal: AbortSignal.timeout(timeoutMs + 10_000), // safety net
      });

      if (res.status === 200) {
        return (await res.json()) as Result;
      }
      if (res.status === 204) {
        // Server timed out — retry with same ID.
        continue;
      }
      if (res.status === 404) {
        throw new RequestNotFoundError(requestId);
      }

      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }

    throw new ResultTimeoutError(requestId, this.maxRetries);
  }

  /**
   * Single-request fast path using `POST /request/sync`.
   *
   * Enqueues the request and waits for the result within the same HTTP
   * connection. Returns `{ result }` when the local server responds before
   * the timeout, or `{ requestId }` when the local server is slow — the
   * caller should then call {@link waitResult} with the returned ID.
   *
   * @throws {QueueFullError} if the queue is at capacity.
   */
  async doSync(
    payload: unknown,
    headers: Record<string, string> = {},
    options?: { timeout?: number }
  ): Promise<{ result: Result } | { requestId: string }> {
    const timeoutMs = options?.timeout ?? this.longPollTimeoutMs;
    const timeoutSec = (timeoutMs / 1000).toFixed(1);
    const url = `${this.baseUrl}/request/sync?timeout=${timeoutSec}s`;

    const res = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Online-API-Key": this.apiKey,
      },
      body: JSON.stringify({ payload, headers }),
      signal: AbortSignal.timeout(timeoutMs + 10_000),
    });

    if (res.status === 200) {
      return { result: (await res.json()) as Result };
    }
    if (res.status === 202) {
      const data = (await res.json()) as { id: string };
      return { requestId: data.id };
    }
    if (res.status === 503) {
      const data = (await res.json()) as { message?: string };
      throw new QueueFullError(data.message);
    }
    throw new HubRouterError(
      `unexpected status ${res.status}: ${await res.text()}`
    );
  }

  /**
   * Submit a request and wait for its result.
   *
   * Tries the single-request fast path ({@link doSync}) first.
   * If the local server is busy, transparently falls back to async polling.
   */
  async do(
    payload: unknown,
    headers: Record<string, string> = {}
  ): Promise<Result> {
    const response = await this.doSync(payload, headers);
    if ("result" in response) {
      return response.result; // fast path — done in one request
    }
    return this.waitResult(response.requestId); // slow path — keep polling
  }
}

File: pkg/client/online.go
// Package client provides ready-to-use clients for both sides of hub-router.
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/jaar23/hub-router/internal/model"
)

// OnlineClient submits requests to hub-router and waits for results.
// It is safe for concurrent use.
type OnlineClient struct {
	baseURL string
	apiKey  string
	opts    OnlineOptions
}

// NewOnlineClient creates an OnlineClient.
// baseURL should be the hub-router base URL, e.g. "https://hub.example.com".
// apiKey is the X-Online-API-Key value.
func NewOnlineClient(baseURL, apiKey string, opts ...func(*OnlineOptions)) *OnlineClient {
	o := defaultOnlineOptions()
	for _, fn := range opts {
		fn(&o)
	}
	return &OnlineClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		apiKey:  apiKey,
		opts:    o,
	}
}

// WithOnlineLongPollTimeout sets the long-poll timeout per request.
func WithOnlineLongPollTimeout(d interface{ String() string }) func(*OnlineOptions) {
	return func(o *OnlineOptions) {
		if dd, ok := d.(interface{ Nanoseconds() int64 }); ok {
			_ = dd
		}
	}
}

// Submit enqueues a request and returns its correlation ID.
// headers is optional metadata forwarded with the request.
func (c *OnlineClient) Submit(ctx context.Context, payload any, headers map[string]string) (string, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("marshal payload: %w", err)
	}

	body := map[string]any{
		"payload": json.RawMessage(payloadBytes),
		"headers": headers,
	}
	bodyBytes, _ := json.Marshal(body)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+"/request", bytes.NewReader(bodyBytes))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Online-API-Key", c.apiKey)

	resp, err := c.opts.HTTPClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("submit request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusServiceUnavailable {
		var errResp model.ErrorResponse
		_ = json.NewDecoder(resp.Body).Decode(&errResp)
		return "", fmt.Errorf("queue full: %s", errResp.Message)
	}
	if resp.StatusCode != http.StatusAccepted {
		raw, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(raw))
	}

	var submitResp model.SubmitResponse
	if err := json.NewDecoder(resp.Body).Decode(&submitResp); err != nil {
		return "", fmt.Errorf("decode submit response: %w", err)
	}
	return submitResp.ID, nil
}

// WaitResult polls GET /result/:id using long-polling until the result arrives
// or MaxRetries 204 responses are received.
// On success returns the Result. On exhaustion returns an error.
func (c *OnlineClient) WaitResult(ctx context.Context, requestID string) (*model.Result, error) {
	timeout := c.opts.LongPollTimeout.String()
	url := fmt.Sprintf("%s/result/%s?timeout=%s", c.baseURL, requestID, timeout)

	for attempt := 0; attempt <= c.opts.MaxRetries; attempt++ {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("X-Online-API-Key", c.apiKey)

		resp, err := c.opts.HTTPClient.Do(req)
		if err != nil {
			return nil, fmt.Errorf("poll result: %w", err)
		}

		switch resp.StatusCode {
		case http.StatusOK:
			var result model.Result
			if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
				resp.Body.Close()
				return nil, fmt.Errorf("decode result: %w", err)
			}
			resp.Body.Close()
			return &result, nil

		case http.StatusNoContent:
			// Server timed out waiting — retry (long-poll is stateless, same ID).
			resp.Body.Close()
			continue

		case http.StatusNotFound:
			resp.Body.Close()
			return nil, fmt.Errorf("request ID %s not found (may have expired)", requestID)

		default:
			raw, _ := io.ReadAll(resp.Body)
			resp.Body.Close()
			return nil, fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(raw))
		}
	}

	return nil, fmt.Errorf("result not available after %d retries for request %s", c.opts.MaxRetries, requestID)
}

// DoSync sends POST /request/sync — a single HTTP request that enqueues and
// waits for the result within the same connection.
//
// Returns:
//   - (result, "", nil)    — fast path: result arrived before timeout.
//   - (nil, requestID, nil) — slow path: local server is busy; caller should
//     call WaitResult(ctx, requestID) to continue polling.
//   - (nil, "", err)       — request could not be submitted (queue full, auth, etc.).
func (c *OnlineClient) DoSync(ctx context.Context, payload any, headers map[string]string) (*model.Result, string, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, "", fmt.Errorf("marshal payload: %w", err)
	}
	body := map[string]any{
		"payload": json.RawMessage(payloadBytes),
		"headers": headers,
	}
	bodyBytes, _ := json.Marshal(body)

	timeout := c.opts.LongPollTimeout.String()
	url := fmt.Sprintf("%s/request/sync?timeout=%s", c.baseURL, timeout)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Online-API-Key", c.apiKey)

	resp, err := c.opts.HTTPClient.Do(req)
	if err != nil {
		return nil, "", fmt.Errorf("sync request: %w", err)
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusOK:
		// Fast path: result came back immediately.
		var result model.Result
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			return nil, "", fmt.Errorf("decode sync result: %w", err)
		}
		return &result, "", nil

	case http.StatusAccepted:
		// Slow path: server timed out waiting; continue with async polling.
		var submitResp model.SubmitResponse
		if err := json.NewDecoder(resp.Body).Decode(&submitResp); err != nil {
			return nil, "", fmt.Errorf("decode submit response: %w", err)
		}
		return nil, submitResp.ID, nil

	case http.StatusServiceUnavailable:
		var errResp model.ErrorResponse
		_ = json.NewDecoder(resp.Body).Decode(&errResp)
		return nil, "", fmt.Errorf("queue full: %s", errResp.Message)

	default:
		raw, _ := io.ReadAll(resp.Body)
		return nil, "", fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(raw))
	}
}

// Do uses DoSync for a single round-trip when the local server is fast,
// falling back to WaitResult for slow processors. Transparent to the caller.
func (c *OnlineClient) Do(ctx context.Context, payload any, headers map[string]string) (*model.Result, error) {
	result, id, err := c.DoSync(ctx, payload, headers)
	if err != nil {
		return nil, err
	}
	if result != nil {
		return result, nil // fast path — done in one request
	}
	return c.WaitResult(ctx, id) // slow path — keep polling
}
